### ISSUE TITLE
Memory grows with the iteration when calling PyTorch

### ISSUE THREAD

**Issue by boathit** [JuliaPy/PyCall.jl#529]
As shown in the following code, the memory grows with the iteration when creating a PyTorch tensor in a for loop but it remains at a constant level if change `X = torch.rand(500, 500)` to `X = np.random[:rand](500, 500)`.
```julia
using PyCall

@pyimport torch
@pyimport numpy as np

for i = 1:10_000
    X = torch.rand(500, 500)
    #X = np.random[:rand](500, 500)
    i % 1000 == 0 && (println("i: $i"); sleep(2))
end
```
Version information: `Julia v0.7` at Ubuntu 16.04 OS.

**Comment by JobJob** [comment]
Probably an extra reference is being held by Julia and torch's garbage collection (gc) code is not being run.

Does this change fix it any chance?
```
for epoch = 1:20
    for (i, (Xs, ys)) in enumerate(trainLoader)
        (Xs, ys) = Xs[:to](device), ys[:to](device)
        i % 100 == 0 && println("Epoch: $epoch")
        gc() # ADD THIS
    end
    #torch.cuda[:empty_cache]()
end
```

**Comment by boathit** [comment]
@JobJob Manual garbage collection `GC.gc()` does not help.

**Comment by Paethon** [comment]
Maybe this is related to #542 I am having with OpenAI Gym?

**Comment by stevengj** [comment]
Hopefully this is fixed by #558.

**Comment by boathit** [comment]
Hi @stevengj, #558 seems to have fixed the problem with `numpy` and `gym`, but the issue remains for `PyTorch`, can you try to run the code that I provide? (I have updated the code to the master branch)

**Comment by JobJob** [comment]
Hi @boathit on PyCall master (w/ Julia 1.0) I can replicate memory growing during the loop, but memory usage goes down again if I call `GC.gc()` after the loop. Can you replicate that on your machine?

I can also trigger the same behaviour with `X = pycall(np.random[:rand], PyObject, 500, 500)` (which avoids the automatic conversion of the returned value to a Julia array).

---
Asides:

I also saw that `typeof(torch.rand(5,5))` is PyObject, which means PyCall isn't able to automatically convert the object returned from Python for `torch.rand(5,5)` to a Julia type, though it can with the result of `np.random[:rand](5,5)`.

(n.b. you can get the raw PyObject with `pycall(torch.rand, PyObject, 5, 5)`)


So a couple of interesting things to note:

1) The performance of the PyObject version is better than the `torch.rand` version, since it avoids the failing attempted conversion to a Julia type (though for large matrix creation the overhead probably won't be significant):
```
using PyCall, BenchmarkTools
@pyimport torch
@pyimport numpy as np

torch_randpyo(args...) = pycall(torch.rand, PyObject, args...)

@btime torch_randpyo(2, 2)  # 1.018 μs (3 allocations: 48 bytes)
@btime torch.rand(2, 2)     # 3.340 μs (6 allocations: 96 bytes)
```

2) The reason Julia can't work out that it's an array is that Torch doesn't support the buffer interface - see a mention of this [here](https://discuss.pytorch.org/t/accessing-the-single-segment-buffer-interface/1354/4) and [here](https://github.com/pytorch/pytorch/issues/9168#issuecomment-402543649). Also in that last link is a [workaround](https://github.com/pytorch/pytorch/issues/9168#issuecomment-402515966) to call `X.numpy()` to convert it to a NumPy array which PyCall knows how to convert to a Julia Array:
```
function torch_randnp(args...)
    Xpyo = torch_randpyo(args...)
    Xpyo["numpy"]()
end

typeof(torch_randnp(2, 2)) # Array{Float32,2}
@btime torch_randnp(2, 2)  # 334.627 μs (78 allocations: 3.00 KiB)
```

The conversion to the Julia array takes a little time, but note that with #487  (wink wink nudge nudge say no more) it goes down to `15.372 μs (51 allocations: 2.52 KiB)`. This is similar to
```
const nprand = np.random["rand"]
typeof(nprand(2, 2)) # Array{Float64,2}
@btime nprand(2, 2)  # 336.512 μs (69 allocations: 2.64 KiB) (12.025 μs with #487)
```

**Comment by boathit** [comment]
@JobJob Manually adding `GC.gc()` does help,  so should this be the problem of Julia instead of PyCall?

**Comment by JobJob** [comment]
@boathit on the Julia v1.0.1 release I'm not really seeing an issue with this, even if I up it to `X = torch.rand(5000, 5000)`. How about you?
---

Aside would recommend this if you want no-copy access to the array
```
X = pycall(torch.rand(500, 500)[:numpy], PyArray) 
i % 1000 == 0 && (println("i: $i, typeof(X): $(typeof(X))"); sleep(2))
```

**Comment by JobJob** [comment]
It's a shame because I had a cute workaround 😄 
```
using PyCall
using MacroTools

@pyimport torch
@pyimport numpy as np

macro pycall!(ex)
    @capture(ex, v_ = fn_(args__))
    esc(quote pycall!($v, $fn, PyObject, $args...) end)
end

const X = PyNULL()
for i = 1:10_000
    @pycall! X = torch.rand(500, 500)
    i % 1000 == 0 && (println("i: $i"); sleep(2))
end
```

**Comment by tkf** [comment]
I've never used pytorch but wouldn't it be better to do `torch.rand(500, 500, out=X)` (where `X` is pre-allocated before the loop)?  Not sure `500, 500` is required in this case though.

https://pytorch.org/docs/stable/torch.html#torch.rand

**Comment by boathit** [comment]
@JobJob The reason I do not want to convert the torch tensor to numpy array is that I need to move the tensor to GPU to perform the subsequent computation, see [this line](https://github.com/boathit/JuliaTorch/blob/master/logisticRegression.jl#L36) in which `device` may be GPU. 

The macro `@pycall!` is cool, but it will break if I move `X` to GPU. 

@tkf `torch.rand(500, 500)` is just a simple case to show the issue. In most of cases, it is unavoidably to generate many intermediate variables that I could not preallocate in advance, see [this line](https://github.com/boathit/JuliaTorch/blob/master/resnet.jl#L102), after each transformation (`self[:f]`) there will be an intermediate tensor (often with different shapes) being generated.

My current solution is explicitly calling `GC.gc()` after a few number of iterations, like [this](https://github.com/boathit/JuliaTorch/blob/master/resnet.jl#L129). This definitely harms the performance, however,  it turns out `PyCall` + `PyTorch` is still the best way to practice deep learning in Julia so far(20x faster than `Flux.jl` running on GPU, the codebase of `Knet.jl` is messy  and I do not want to hack it anymore 😂).

**Comment by JobJob** [comment]
Do you still need to call GC.gc() on Julia v1.0.1? I found that I didn't need to anymore.

You could also potentially call `Base.finalize(x)` explicitly, instead of GC.gc() which may be faster(though not sure if that is recommended either). At some point soon I'll hopefully get a chance to take a closer look at your examples.

> PyCall + PyTorch is still the best way to practice deep learning in Julia so far(20x faster than Flux.jl running on GPU

Interesting, do you have any decent benchmark code comparing Flux and PyTorch? Would be good to post it at Flux.jl - 20x is surprising

**Comment by tkf** [comment]
Right, preallocation doesn't work well with define-by-run...

If you want to avoid GC overhead, I think you'd better avoid Python-Julia round trip in the first place. Using `pybuiltin("exec")` may actually be better in terms of performance. Or define a Python function and call it.

**Comment by boathit** [comment]
@JobJob It is weird, I also use Julia v1.0.1, PyCall.jl 1.18.5 on Ubuntu 14.04 and I can confirm both CPU memory and GPU memory will grow with the iteration, see the `used` column below.

```
~/JuliaTorch » free -h                                                                                            xiucheng@gpu8
             total       used       free     shared    buffers     cached
Mem:          125G        15G       109G       156M       1.1G        11G
-/+ buffers/cache:       3.5G       122G
Swap:         119G         0B       119G
------------------------------------------------------------
~/JuliaTorch » free -h                                                                                            xiucheng@gpu8
             total       used       free     shared    buffers     cached
Mem:          125G        17G       108G       156M       1.1G        11G
-/+ buffers/cache:       5.2G       120G
Swap:         119G         0B       119G
------------------------------------------------------------
~/JuliaTorch » free -h                                                                                            xiucheng@gpu8
             total       used       free     shared    buffers     cached
Mem:          125G        18G       107G       156M       1.1G        11G
-/+ buffers/cache:       5.6G       120G
Swap:         119G         0B       119G
------------------------------------------------------------
~/JuliaTorch » free -h                                                                                            xiucheng@gpu8
             total       used       free     shared    buffers     cached
Mem:          125G        20G       105G       156M       1.1G        11G
-/+ buffers/cache:       7.6G       118G
Swap:         119G         0B       119G
------------------------------------------------------------
~/JuliaTorch » free -h                                                                                            xiucheng@gpu8
             total       used       free     shared    buffers     cached
Mem:          125G        22G       103G       156M       1.1G        11G
-/+ buffers/cache:        10G       115G
Swap:         119G         0B       119G
------------------------------------------------------------
```


Maybe I should try PyCall.jl on the master branch.

As for the benchmark, I will make one soon.

**Comment by boathit** [comment]
> Using pybuiltin("exec") may actually be better in terms of performance. Or define a Python function and call it.

@tkf, Can you be more specific about that? It will be a great help if you can give me a simple example as I am not that familiar with PyCall.jl. Thanks.

**Comment by JobJob** [comment]
Yeah, CPU memory grows for me for a bit, but then it clears without an explicit call to `GC.gc()` (it gets to a higher peak with (500, 500) than (5000, 5000) - say 7GB on my system), but yeah that's a lot of GBs.

I'm on mac though, haven't checked GPU mem.

**Comment by tkf** [comment]
There is no memory-related fix since v1.18.5: https://github.com/JuliaPy/PyCall.jl/compare/v1.18.5...master

I was talking about something like this:

```julia
julia> ns = PyDict(Dict(:a=>1, :b=>2, :c=>3))
       pybuiltin("exec")("""
       x = a
       y = x + b
       z = y + c
       ans = z
       """, ns)

julia> ns[:ans]
6
```

If you create a lot of intermediate Python variables (like x, y, z above) inside the Python code and you only need to pull out a few results from the computation, it makes sense to do this since those intermediate variables would not be brought into the Julia namespace so won't pressure Julia's GC.

But you'd use PyTorch as-is if you are fine with writing Python in the first place. So this probably does not make any sense :)

**Comment by tkf** [comment]
I just noticed that `GC.gc` is defined by

```julia
gc(full::Bool=true) = ccall(:jl_gc_collect, Cvoid, (Int32,), full)
```

I have no idea `full` argument is but passing `false` sounds like it would do less amount of work :laughing:

**Comment by boathit** [comment]
@tkf Oh, I see. The main motivation I wish to use PyTorch in Julia is that, in many machine learning models, it may involve a large number of iterated operations such as sampling, which will be slow in Python. In that case, I still have to pull out those variables into Julia.

**Comment by tkf** [comment]
It's interesting that you can beat Python even with the conversion (and GC) overhead.

**Comment by boathit** [comment]
Yeah, but I can also confirm that `GC.gc(false)` is able to bring down the overhead significantly (in comparison to `GC.gc()`) and at the same time the memory usage is kept at a constant level.

**Comment by boathit** [comment]
@JobJob It seems both Flux and CuArrays have undergone significant advances since I tested them last time, that's what I would love to see. I made a benchmark [here](https://github.com/boathit/Benchmark-Flux-PyTorch). At present Flux is at least faster in the first two test cases. As Flux's `BatchNorm` has crashed on GPU right now, I have to wait to fill its residual network part later, and its RNN will also be added.
